src/background.rs

```diff
@@ -0,0 +1,489 @@
+
+use influxdb::{Client, WriteQuery};
+use tokio::sync::{mpsc, oneshot};
+
+use tokio_stream::{wrappers::UnboundedReceiverStream, StreamExt};
+
+use crate::{
+    errors::Error,
+    events::{Event, EventType},
+    network_conf::NetworkConfiguration,
+    params::RunParameters,
+    requests::{PlayloadType, Request, RequestType},
+    responses::{Response, ResponseType},
+    websocket::WebsocketClient,
+};
+
+const WEBSOCKET_RECEIVER: &str = "Websocket Receiver";
+const WEBSOCKET_SENDER: &str = "Websocket Sender";
+
+#[derive(Debug)]
+pub enum Command {
+    Publish {
+        topic: String,
+        message: String,
+        sender: oneshot::Sender<Result<u64, Error>>,
+    },
+    Subscribe {
+        topic: String,
+        stream: mpsc::UnboundedSender<Result<String, Error>>,
+    },
+
+    SignalEntry {
+        state: String,
+        sender: oneshot::Sender<Result<u64, Error>>,
+    },
+
+    Barrier {
+        state: String,
+        target: u64,
+        sender: oneshot::Sender<Result<(), Error>>,
+    },
+
+    WaitNetworkInitializedStart {
+        sender: oneshot::Sender<Result<u64, Error>>,
+    },
+
+    WaitNetworkInitializedBarrier {
+        sender: oneshot::Sender<Result<(), Error>>,
+    },
+
+    WaitNetworkInitializedEnd {
+        sender: oneshot::Sender<Result<u64, Error>>,
+    },
+
+    NetworkShaping {
+        config: NetworkConfiguration,
+        sender: oneshot::Sender<Result<u64, Error>>,
+    },
+
+    SignalSuccess {
+        sender: oneshot::Sender<Result<u64, Error>>,
+    },
+
+    SignalFailure {
+        error: String,
+        sender: oneshot::Sender<Result<u64, Error>>,
+    },
+
+    SignalCrash {
+        error: String,
+        stacktrace: String,
+        sender: oneshot::Sender<Result<u64, Error>>,
+    },
+
+    Metric {
+        write_query: WriteQuery,
+        sender: oneshot::Sender<Result<(), Error>>,
+    },
+}
+
+#[derive(Debug)]
+enum PendingRequest {
+    PublishOrSignal {
+        sender: oneshot::Sender<Result<u64, Error>>,
+    },
+    Barrier {
+        sender: oneshot::Sender<Result<(), Error>>,
+    },
+    Subscribe {
+        stream: mpsc::UnboundedSender<Result<String, Error>>,
+    },
+}
+
+pub struct BackgroundTask {
+    websocket_tx: mpsc::UnboundedSender<(Request, oneshot::Sender<Result<(), Error>>)>,
+    websocket_rx: UnboundedReceiverStream<Result<Response, Error>>,
+
+    influxdb: influxdb::Client,
+
+    next_id: u64,
+
+    params: RunParameters,
+
+    client_rx: UnboundedReceiverStream<Command>,
+
+    pending_req: HashMap<u64, PendingRequest>,
+}
+
+impl BackgroundTask {
+    pub async fn new(
+        client_rx: mpsc::UnboundedReceiver<Command>,
+        params: RunParameters,
+    ) -> Result<Self, Box<dyn std::error::Error>> {
+        let (websocket_tx, req_rx) = mpsc::unbounded_channel();
+        let (res_tx, websocket_rx) = mpsc::unbounded_channel();
+
+        let web = WebsocketClient::new(res_tx, req_rx).await?;
+
+        tokio::spawn(web.run());
```

Comment:
@[mxinden](https://github.com/mxinden) mxinden [yesterday](https://github.com/testground/sdk-rust/pull/4#discussion_r827974432)

Is spawning the websocket work onto a separate task necessary? If I am not mistaken, Commands like Command::Publish await the flushing of the message by the separate task via the provided oneshot anyways, thus making this a synchronous operation. Would it not be simpler to do all of this in a single task?

I doubt we need to optimize for performance on any of this.

---


@[SionoiS](https://github.com/SionoiS) SionoiS [yesterday](https://github.com/testground/sdk-rust/pull/4#discussion_r828222519)

I failed at combining the 2. It's surely possible but my async knownlege is not there yet.


---

@SionoiS I can propose a patch against this pull request if you like.

Reply:
Can't wait to see what you come up with!